File: MilkVisits.py
class node:
    def __init__(self,b):
        self.t=b
        self.c=[self]
    def addToPath(self,n):
        x = self.c
        self.c.append(n)
        for i in n.c:
            self.c.append(i)
        for i in x:
            i.c.append(n)
        return
fin = open("milkvisits.in","r")
(n,m) = (int(i) for i in fin.readline().split(" "))
s = fin.readline()
c=[]
for i in range(n):
    c.append(node((s[i]=="H")))
for i in range(n-1):
    (x,y)=(int(i)-1 for i in fin.readline().split(" "))
    if(c[x].t==c[y].t):
        c[x].addToPath(c[y])
        c[y].addToPath(c[x])
o=""
for i in range(m):
    (ix,iy,it) = (i for i in fin.readline().split(" "))
    x=c[int(ix)-1]
    y=c[int(iy)-1]
    t = (it=="H\n")
    if (y in x.c) and (t!=x.t):
        o+="0"
    else:
        o+="1"
fin.close()
fout = open("milkvisits.out","w+")
fout.write(o+"\n")
fout.close()
print(o)

File: MilkVisits.cpp
#include <iostream>
#include <fstream>
#include <vector>
#include <algorithm>
#include <string>
using namespace std;
class node
{
public:
    vector<node> c;
    int n;
    bool t;
    node(int i, bool b)
    {
        n = i;
        t = b;
        c.push_back(*this);
    }
    void addToPath(node n)
    {
        c.push_back(n);
        for (int i = 0; i < n.c.size(); i++)
        {
            c.push_back(n.c[i]);
        }
        for (int i = 0; i < c.size() - n.c.size() - 1; i++)
        {
            c[i].c.push_back(n);
            for (int i = 0; i < n.c.size(); i++)
            {   
                c[i].c.push_back(n.c[i]);
            }
        }
    }
    bool connectedTo(int x)
    {
        for (int i = 0; i < c.size(); i++)
        {
            if (c[i].n == x)
            {
                return (true);
            }
        }
        return (false);
    }
};
int main()
{
    ifstream in("milkvisits.in");
    int n, m;
    in >> n >> m;
    char s = ' ';
    vector<node> c;
    for (int i = 0; i < n; i++)
    {
        in >> s;
        c.push_back(node(i+1, (s == 'H')));
    }
    for (int i = 0; i < n - 1; i++)
    {
        int x, y;
        in >> x >> y;
        x--;
        y--;
        if (c[x].t ^ c[y].t)
        {
            c[x].addToPath(c[y]);
            c[y].addToPath(c[x]);
        }
    }
    string o;
    for (int i = 0; i < m; i++)
    {
        int x, y;
        char t;
        in >> x >> y >> t;
        x--;
        bool b = (t == 'H');
        if (c[x].connectedTo(y) && (c[x].t ^ b))
        {
            o += "0";
        }
        else
        {
            o += "1";
        }
        cout << t << endl;
    }
    cout << o;
    in.close();
    ofstream out("milkvisits.out");
    out << o << endl;
    out.close();
    return (0);
}